data_server/main.py

```diff
@@ -35,5 +37,27 @@ def generate_response(data: bytes):
                     headers=headers)
 
 
+@app.route('/dataset', methods=['GET'])
+def get_dataset():
+    """Downloads and returns the requested dataset if it exists."""
+    dataset_name = request.args.get('name')
+    if dataset_name is None:
+        return 'Request missing required url param \'name\'', 400
+
+    try:
+        dataset = gcs_utils.download_blob_as_bytes(
+            os.environ.get('GCS_BUCKET'), dataset_name)
+    except google.cloud.exceptions.NotFound:
+        return 'Dataset {} not found'.format(dataset_name), 404
+
+    def generate_response(data: bytes):
+        for row in dataset.splitlines():
+            yield row + b'\n'
+    headers = Headers()
+    headers.add('Content-Disposition', 'attachment', filename=dataset_name)
```

Comment:
You'll need to add
```
headers.add('Access-Control-Allow-Origin', '*')
headers.add('Vary', 'Origin ')
```
to allow the frontend to request the data

Reply:
Done. Do you have a link that explains these headers? I'm not very familiar with them.

Reply:
Yep, I'd start here: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS

TL;DR is that because the data server and our frontend are on different origins, the frontend cannot read the contents of responses from the data server unless the data server provides the "Access-Control-Allow-Origin" header with a value that matches the frontend's origin. Passing the "*" value says that any origin can read it (which I think is fine per offline discussions about not using CORS as a privacy mechanism and that it's fine if the API is public). The "Vary" header specifies how intermediate caches (like CDNs, or I think even local browser http cache) uniquely identify a resource. So specifying "Origin" means that separate origins will not share a cache. This is important to make sure the access control is respected even from cache. I'm guessing flask doesn't add caching headers by default so the Vary header probably doesn't matter, but it's good practice to be safe.

Actually, I forgot - the Vary header value should be `Accept-Encoding, Origin`. The Accept-Encoding piece makes sure intermediate caches don't serve the wrong encoding. Here's a good explanation: https://blog.stackpath.com/accept-encoding-vary-important/

Reply:
Thanks! These are really helpful